src/NSubstitute/Proxies/DelegateProxy/DelegateProxyFactory.cs

```diff
@@ -1,28 +1,38 @@
 using System;
-using System.Collections.Generic;
+using System.Collections.Concurrent;
+using System.Globalization;
 using System.Linq;
-using System.Linq.Expressions;
 using System.Reflection;
+using System.Reflection.Emit;
+using System.Threading;
 using NSubstitute.Core;
 using NSubstitute.Exceptions;
 
 namespace NSubstitute.Proxies.DelegateProxy
 {
     public class DelegateProxyFactory : IProxyFactory
     {
-        private readonly IArgumentSpecificationDequeue _argSpecificationDequeue;
+        private const string MethodNameInsideProxyContainer = "Invoke";
+        private readonly IProxyFactory _objectProxyFactory;
+        private readonly ModuleBuilder _moduleBuilder;
+        private readonly ConcurrentDictionary<Type, Type> _delegateContainerCache = new ConcurrentDictionary<Type, Type>();
+        private long _typeSuffixCounter;
 
-        public DelegateProxyFactory(IArgumentSpecificationDequeue argSpecificationDequeue)
+        public DelegateProxyFactory(IProxyFactory objectProxyFactory)
         {
-            _argSpecificationDequeue = argSpecificationDequeue;
+            _objectProxyFactory = objectProxyFactory;
+            _moduleBuilder = AssemblyBuilder
+                .DefineDynamicAssembly(new AssemblyName("NSubsituteDelegateProxyTypes"), AssemblyBuilderAccess.Run)
+                .DefineDynamicModule("NSubsituteDelegateProxyTypes");
```

Comment:
Suggest pulling out "NSubsituteDelegateProxyTypes" as a constant to avoid these line 25 and 26 accidentally getting out of sync.